### PR TITLE
Provide access to verify_peer_callback from C#

### PR DESCRIPTION
**Pull request by tlg-bf**
At the moment there is no possibilities from C# to access the server certificate from the client side when using TLS. In P2P settings it is practical to be able to identify a node in the network using its TLS certificate.
This PR exposes `verify_peer_callback` available in the C++ layer to the C# library to be able to obtain the server certificate.

Notes:
* I defined two types of delegates `VerifyPeerCallback` and `VerifyCallbackInternal`. The first is the one exposed to be used externally of the library, the second is the one passed to the native code. The reason for defining both was to hide the `userData` parameter of the callback (it is not clear to me if/how keeping this parameter could work).
* I added an optional parameter to the `SslCredentials` constructors to pass a `VerifyPeerCallback` delegate.
* I added some tests in "src/csharp/Grpc.IntegrationTesting/SslCredentialsTest.cs" to test the proper working of the functionality.
* "src/csharp/ext/grpc_csharp_ext.c" was modified so that `grpcsharp_ssl_credentials_create` takes the extra delegate parameter.

### CODE REVIEW

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @tlg-bf isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Comment by tlg-bf** [comment]
Should be covered by CLA now.

**Comment by tlg-bf** [comment]
@jtattermusch Sorry to bother, but would there be any chance on getting a feedback on this PR? We would really like to switch to GRPC but not being able to access the identity of the peer really makes it impossible at the moment. If the approach I have taken is not proper or if there is a better way of doing it I would be more than happy to update.

File: src/csharp/Grpc.Core/ChannelCredentials.cs [hunk 1]
```diff
@@ -125,19 +142,23 @@ public SslCredentials() : this(null, null)
         /// Creates client-side SSL credentials from
         /// a string containing PEM encoded root certificates.
         /// </summary>
-        public SslCredentials(string rootCertificates) : this(rootCertificates, null)
+        public SslCredentials(string rootCertificates, VerifyPeerCallback verifyPeerCallback = null) :
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/ChannelCredentials.cs]: adding a new default argument breaks binary compatibility. If we want to add a new argument, we need to add a new constructor.

File: src/csharp/Grpc.Core/ChannelCredentials.cs [hunk 2]
```diff
@@ -125,19 +142,23 @@ public SslCredentials() : this(null, null)
...
-            
+
         /// <summary>
         /// Creates client-side SSL credentials.
         /// </summary>
         /// <param name="rootCertificates">string containing PEM encoded server root certificates.</param>
         /// <param name="keyCertificatePair">a key certificate pair.</param>
-        public SslCredentials(string rootCertificates, KeyCertificatePair keyCertificatePair)
+        /// <param name="verifyPeerCallback">a callback to verify peer's target name and certificate.</param>
+        public SslCredentials(string rootCertificates, KeyCertificatePair keyCertificatePair, VerifyPeerCallback verifyPeerCallback = null)
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/ChannelCredentials.cs]: let's just add one new constructor 
`SslCredentials(string rootCertificates, KeyCertificatePair keyCertificatePair, VerifyPeerCallback verifyPeerCallback)`

File: src/csharp/Grpc.Core/ChannelCredentials.cs [hunk 3]
```diff
@@ -104,13 +104,30 @@ internal override ChannelCredentialsSafeHandle CreateNativeCredentials()
...
+    /// <summary>
+    /// Callback invoked with the expected targetHost and the peer's certificate.
+    /// If a non-zero value is returned by this callback then it is treated as a
+    /// verification failure. Invocation of the callback is blocking, so any
+    /// implementation should be light-weight.
+    /// </summary>
+    /// <param name="targetHost">string containing the host name of the peer.</param>
+    /// <param name="targetPem">string containint PEM encoded certificate of the peer.</param>
+    /// <returns>0 is verification succeeded, non 0 otherwise.</returns>
+    public delegate int VerifyPeerCallback(string targetHost, string targetPem);
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/ChannelCredentials.cs]: return `bool` instead of `int`

File: src/csharp/Grpc.Core/ChannelCredentials.cs (hunk 1 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/ChannelCredentials.cs]: Removed default argument and added a new constructor.

File: src/csharp/Grpc.Core/ChannelCredentials.cs (hunk 2 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/ChannelCredentials.cs]: Added

File: src/csharp/Grpc.Core/ChannelCredentials.cs (hunk 3 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/ChannelCredentials.cs]: Replaced

**Comment by tlg-bf** [comment]
Thank you for the comments. I tried to addressed them, let me know if there is anything more that should be done.

I also rebased on master and fixed some issues with the tests that I had added (I thought they were working before but trying again seemed initialization was failing).

File: src/csharp/Grpc.Core/ChannelCredentials.cs [hunk 4]
```diff
@@ -104,40 +104,71 @@ internal override ChannelCredentialsSafeHandle CreateNativeCredentials()
...
         /// <param name="rootCertificates">string containing PEM encoded server root certificates.</param>
         /// <param name="keyCertificatePair">a key certificate pair.</param>
-        public SslCredentials(string rootCertificates, KeyCertificatePair keyCertificatePair)
+        /// <param name="verifyPeerCallback">a callback to verify peer's target name and certificate.</param>
+        public SslCredentials(string rootCertificates, KeyCertificatePair keyCertificatePair, VerifyPeerCallback verifyPeerCallback)
         {
             this.rootCertificates = rootCertificates;
             this.keyCertificatePair = keyCertificatePair;
+            this.verifyPeerCallback = verifyPeerCallback == null ?
+                null : new VerifyPeerCallbackInternal((host, pem, userData) => verifyPeerCallback(host, pem) ? 0 : 1);
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/ChannelCredentials.cs]: this is actually not safe - the problem is that the delegate can be garbage collected at some point but the native layer won't know about it and will try to invoke it, which will result in segfault.
Native to managed callbacks are tricky.
Look at https://github.com/grpc/grpc/blob/3278bdceda8030d5aa130f12765e5f07263c860d/src/csharp/Grpc.Core/Internal/NativeMetadataCredentialsPlugin.cs#L47 to see how the destruction logic is handled (it's not that simple).

File: src/csharp/Grpc.Core/ChannelCredentials.cs [hunk 5]
```diff
@@ -104,40 +104,71 @@ internal override ChannelCredentialsSafeHandle CreateNativeCredentials()
...
     }
 
+    /// <summary>
+    /// Callback invoked with the expected targetHost and the peer's certificate.
+    /// If a non-zero value is returned by this callback then it is treated as a
+    /// verification failure. Invocation of the callback is blocking, so any
+    /// implementation should be light-weight.
+    /// </summary>
+    /// <param name="targetHost">string containing the host name of the peer.</param>
+    /// <param name="targetPem">string containint PEM encoded certificate of the peer.</param>
```
[the remark this answers is not in the document]
Comment (tlg-bf) [reply src/csharp/Grpc.Core/ChannelCredentials.cs]: No longer relevant after changes to `VerifyPeerContext`.

File: src/csharp/Grpc.Core/ChannelCredentials.cs [hunk 6]
```diff
@@ -104,40 +104,71 @@ internal override ChannelCredentialsSafeHandle CreateNativeCredentials()
...
+    /// <returns>0 is verification succeeded, non 0 otherwise.</returns>
```
[the remark this answers is not in the document]
Comment (tlg-bf) [reply src/csharp/Grpc.Core/ChannelCredentials.cs]: Updated.

File: src/csharp/Grpc.Core/ChannelCredentials.cs [hunk 7]
```diff
@@ -104,40 +104,71 @@ internal override ChannelCredentialsSafeHandle CreateNativeCredentials()
...
+    public delegate bool VerifyPeerCallback(string targetHost, string targetPem);
```
[the remark this answers is not in the document]
Comment (tlg-bf) [reply src/csharp/Grpc.Core/ChannelCredentials.cs]: Indeed makes sense, implemented as proposed.

File: src/csharp/Grpc.Core/ChannelCredentials.cs [hunk 8]
```diff
@@ -104,40 +104,71 @@ internal override ChannelCredentialsSafeHandle CreateNativeCredentials()
...
+
+    internal delegate int VerifyPeerCallbackInternal(
```
[the remark this answers is not in the document]
Comment (tlg-bf) [reply src/csharp/Grpc.Core/ChannelCredentials.cs]: Moved the delegate definition to `ChannelCredentialsSafeHandle.cs`.

File: src/csharp/Grpc.Core/ChannelCredentials.cs [hunk 9]
```diff
@@ -104,40 +104,71 @@ internal override ChannelCredentialsSafeHandle CreateNativeCredentials()
...
+        }
+
         /// <summary>
         /// Creates client-side SSL credentials.
         /// </summary>
         /// <param name="rootCertificates">string containing PEM encoded server root certificates.</param>
         /// <param name="keyCertificatePair">a key certificate pair.</param>
-        public SslCredentials(string rootCertificates, KeyCertificatePair keyCertificatePair)
+        /// <param name="verifyPeerCallback">a callback to verify peer's target name and certificate.</param>
+        public SslCredentials(string rootCertificates, KeyCertificatePair keyCertificatePair, VerifyPeerCallback verifyPeerCallback)
```
[the remark this answers is not in the document]
Comment (tlg-bf) [reply src/csharp/Grpc.Core/ChannelCredentials.cs]: Added this note to a couple of places where I thought it made sense:
* The `VerifyPeerCallback` delegate definition (https://github.com/grpc/grpc/pull/17051/files#diff-4281d7ff0b726dd27c1d9f64ee85fca6R128).
* This constructor (https://github.com/grpc/grpc/pull/17051/files#diff-4281d7ff0b726dd27c1d9f64ee85fca6R174).
* The `VerifyPeerContext` class (https://github.com/grpc/grpc/pull/17051/files#diff-9a333bdf5d97c42cae8e54ecae812730R5).

File: src/csharp/Grpc.Core/ChannelCredentials.cs (hunk 4 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/ChannelCredentials.cs]: Thank you for pointing me to this. I had sort of guessed that there could be some issue with the callback. I first added a `GC.Collect()` in one of the test to highlight the issue (https://github.com/grpc/grpc/pull/17051/files#diff-576c954e2167da66d782732e2a4196d7R226) and then implemented a fix for it. I would have like the fix to be less impactful but could not really find a better way than what I have done now. The main issue is that the channel does not keep track of the channel credentials that are passed in its constructor. A naive way would have been to simply add a reference to it, but I thought that only keeping track of the `GCHandle` when necessary would be slightly cleaner. Let me know what you think.

File: src/csharp/Grpc.Core/Channel.cs [hunk 10]
```diff
@@ -84,6 +85,13 @@ public Channel(string target, ChannelCredentials credentials, IEnumerable<Channe
                     this.handle = ChannelSafeHandle.CreateInsecure(target, nativeChannelArgs);
                 }
             }
+
+            if (credentials.HasGcHandle)
+            {
+                var gcHandle = credentials.GCHandle;
+                cleanUpCredentials = () => gcHandle.Free();
```
Comment (tlg-bf) [inline src/csharp/Grpc.Core/Channel.cs]: The reason for using an `Action` instead of just a `GCHandle` field is simply to avoid creating and `GCHandle` when it is not needed. It would not have a noticeable impact but just thought that this approach was a tiny bit cleaner.

**Comment by AspirinSJL** [comment]
@jtattermusch Friendly reminder.

File: src/csharp/Grpc.Core/Channel.cs [hunk 11]
```diff
@@ -252,6 +260,7 @@ public async Task ShutdownAsync()
             lock (myLock)
             {
                 handle.Dispose();
+                cleanUpCredentials?.Invoke();
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Channel.cs]: this is not the right time to cleanup the native callback.

verify_peer_destruct  from verify_peer_options should provide the signal to free the gchandle.
refer to https://github.com/grpc/grpc/blob/b3b5d634231ce2c5c0ec0c557b6844e1a43b482e/src/csharp/Grpc.Core/Internal/NativeMetadataCredentialsPlugin.cs#L57  for an example.
IMHO there should be no changes in Channel.cs at all necessary.

File: src/csharp/Grpc.Core/ChannelCredentials.cs [hunk 12]
```diff
@@ -169,9 +212,36 @@ internal override bool IsComposable
             get { return true; }
         }
 
+        internal override bool HasGcHandle
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/ChannelCredentials.cs]: there seems to be too much unnecessary boilerplate around the gcHandle, the two newly added properties don't add any value.

File: src/csharp/Grpc.Core/Channel.cs (hunk 10 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/Channel.cs]: No longer relevant on latest verison.

File: src/csharp/Grpc.Core/Channel.cs (hunk 11 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/Channel.cs]: Indeed using `verify_peer_destruct` is much better, hadn't thought of that. Updated to use it.

File: src/csharp/Grpc.Core/ChannelCredentials.cs (hunk 12 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/ChannelCredentials.cs]: All of this is now removed.

File: src/csharp/Grpc.Core/Channel.cs (hunk 11 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/Channel.cs]: Actually realized that I also needed a gcHandle for the destruct callback. But looking at it, I must say I am having trouble to understand how things work, even for the example your provided.

In `NativeMetadataCredentialsPlugin`, the `NativeMetadataInterceptorHandler` references `gcHandle` which is a field in the class. But if an instance of the class gets GC'd, wouldn't it be collected as well? Or does it mean that the entire object will not be collected until the `GCHandle` is realeased?

I tried to look for documentation but could not find much, I just want to make sure that my implementation makes sense.

Comment (jtattermusch) [reply src/csharp/Grpc.Core/Channel.cs]: In NativeMetadataCredentialsPlugin I'm actually squashing more native callbacks into just one with more parameters. 

Ad releasing:   gcHandle is on the delegate object, but the delegate implicitly reference and instance of the class ("this") so the class can't get garbage collected as long as the gcHandle is allocated. (I think referencing the class instance itself in the GCHandle would work too).

File: src/csharp/Grpc.Core/ChannelCredentials.cs [hunk 13]
```diff
@@ -104,40 +105,74 @@ internal override ChannelCredentialsSafeHandle CreateNativeCredentials()
...
     /// <summary>
     /// Client-side SSL credentials.
     /// </summary>
     public sealed class SslCredentials : ChannelCredentials
     {
         readonly string rootCertificates;
         readonly KeyCertificatePair keyCertificatePair;
+        readonly VerifyPeerCallbackInternal verifyPeerCallback;
+        readonly VerifyPeerDestruct verifyPeerDestruct;
+        readonly GCHandle gcHandle;
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/ChannelCredentials.cs]: you don't need two GCHandle. 
- either you can use just one native callback (with more params)
- or you can alloc GCHandle on the SslCredentials class instance itself (the instance references the delegates so they cannot be GC'd).

File: src/csharp/Grpc.Core/Internal/ChannelCredentialsSafeHandle.cs [hunk 14]
```diff
@@ -15,11 +15,16 @@
 #endregion
 using System;
 using System.Runtime.InteropServices;
-using System.Threading;
-using System.Threading.Tasks;
 
 namespace Grpc.Core.Internal
 {
+    internal delegate int VerifyPeerCallbackInternal(
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/ChannelCredentialsSafeHandle.cs]: just use one delegate  with all the args and `isDestroy` bool flag to simplify things?

Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/ChannelCredentialsSafeHandle.cs]: I think the targetHost and targetPem need to be IntPtr and you need to convert them to a string yourself (again see my example). the problem is that the automatic char* to string conversion might do copying or deallocation of the char* and that's something you don't want.

https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Core/Internal/NativeMetadataCredentialsPlugin.cs#L65

Comment (tlg-bf) [reply src/csharp/Grpc.Core/Internal/ChannelCredentialsSafeHandle.cs]: Yes you're right thanks, fixed it.

Comment (tlg-bf) [reply src/csharp/Grpc.Core/Internal/ChannelCredentialsSafeHandle.cs]: Done. There was no `state` object like in the example you provided me, so I made use of the `userdata` pointer in the native code to keep track of the callback, I hope that's fine.

File: src/csharp/Grpc.Core/ChannelCredentials.cs (hunk 13 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/ChannelCredentials.cs]: Switched to a single native callback.

File: src/csharp/Grpc.Core/Channel.cs (hunk 11 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/Channel.cs]: I hadn't looked at the native part for your example, sorry for that, I think I got it now.

Thanks for the answer regarding `gcHandle`. However I'm still a bit curious, doing some reading on the subject, I found [this article ](https://www.microsoftpressstore.com/articles/article.aspx?p=2224054&seqNum=4) that seems to say that you're supposed to pass the `gcHandle` to the native code and not the object you want to callback on in case of the object being moved by memory compaction during GC:
> You can’t actually pass a pointer to a managed object out to native code, because if a garbage collection occurs, the object could move in memory, invalidating the pointer. So to work around this, you would call GCHandle’s Alloc method, passing in a reference to the object and the Normal flag. Then you’d cast the returned GCHandle instance to an IntPtr and pass the IntPtr into the native code.

Is that not an issue in this case?

Comment (jtattermusch) [reply src/csharp/Grpc.Core/Channel.cs]: I haven't read it in full, but it seems to talk about a different case (when you're passing a pointer to managed object to the native code and the native code directly accesses the representation of that object in memory). We're not doing that, we're only passing a delegate and the marshalling semantics of P/Invoke in that case is to translate the delegate into a native callback and pass that one to the native code (this transitions object is called the "thunk" in the docs btw).

File: src/csharp/Grpc.IntegrationTesting/SslCredentialsTest.cs [hunk 15]
```diff
@@ -44,17 +44,23 @@ public class SslCredentialsTest
...
+        List<ChannelOption> options;
+        bool isHostEqual;
+        bool isPemEqual;
 
         public void InitClientAndServer(bool clientAddKeyCertPair,
                 SslClientCertificateRequestType clientCertRequestType)
         {
             rootCert = File.ReadAllText(TestCredentials.ClientCertAuthorityPath);
+            certChain = File.ReadAllText(TestCredentials.ServerCertChainPath);
+            certChain = certChain.Replace("\r", string.Empty);
```
Comment (jtattermusch) [inline src/csharp/Grpc.IntegrationTesting/SslCredentialsTest.cs]: why this change?

File: src/csharp/Grpc.Core/ChannelCredentials.cs [hunk 16]
```diff
@@ -104,40 +105,72 @@ internal override ChannelCredentialsSafeHandle CreateNativeCredentials()
         }
     }
 
+    /// <summary>
+    /// Callback invoked with the expected targetHost and the peer's certificate.
+    /// If a non-zero value is returned by this callback then it is treated as a
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/ChannelCredentials.cs]: "non-zero"  makes no sense here, it returns bool.

File: src/csharp/Grpc.Core/ChannelCredentials.cs [hunk 17]
```diff
@@ -171,7 +204,20 @@ internal override bool IsComposable
 
         internal override ChannelCredentialsSafeHandle CreateNativeCredentials()
         {
-            return ChannelCredentialsSafeHandle.CreateSslCredentials(rootCertificates, keyCertificatePair);
+            return ChannelCredentialsSafeHandle.CreateSslCredentials(rootCertificates, keyCertificatePair, this.verifyPeerCallbackInternal);
+        }
+
+        private int VerifyPeerCallbackInternal(IntPtr host, IntPtr pem, IntPtr userData, bool isDestroy)
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/ChannelCredentials.cs]: nit: rename to VerifyPeerCallbackHandler to avoid confusion with the `verifyPeerCallbackInternal` member?

File: src/csharp/Grpc.Core/ChannelCredentials.cs [hunk 18]
```diff
@@ -171,7 +204,20 @@ internal override bool IsComposable
...
+        {
+            if (isDestroy)
+            {
+                this.gcHandle.Free();
+                return 0;
+            }
+
+            var context = new VerifyPeerContext(Marshal.PtrToStringAnsi(host), Marshal.PtrToStringAnsi(pem));
+
+            return this.verifyPeerCallback(context) ? 0 : 1;
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/ChannelCredentials.cs]: you need to catch exceptions here, throwing from within a native callback will crash the application.
See https://github.com/grpc/grpc/blob/8fd43c2e91d6fb4f2cd62b82ec1fe7b2218cd66f/src/csharp/Grpc.Core/Internal/NativeMetadataCredentialsPlugin.cs#L71

File: src/csharp/Grpc.Core/Internal/NativeMethods.Generated.cs [hunk 19]
```diff
@@ -478,7 +478,7 @@ public class Delegates
             public delegate void grpcsharp_channel_args_set_integer_delegate(ChannelArgsSafeHandle args, UIntPtr index, string key, int value);
             public delegate void grpcsharp_channel_args_destroy_delegate(IntPtr args);
             public delegate void grpcsharp_override_default_ssl_roots_delegate(string pemRootCerts);
-            public delegate ChannelCredentialsSafeHandle grpcsharp_ssl_credentials_create_delegate(string pemRootCerts, string keyCertPairCertChain, string keyCertPairPrivateKey);
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/NativeMethods.Generated.cs]: this file is generated - you need to edit the template and run `tools/buildgen/generate_projects.sh`

File: src/csharp/ext/grpc_csharp_ext.c [hunk 20]
```diff
@@ -927,20 +927,53 @@ grpcsharp_override_default_ssl_roots(const char* pem_root_certs) {
   grpc_set_ssl_roots_override_callback(override_ssl_roots_handler);
 }
 
+typedef int(GPR_CALLTYPE* grpcsharp_verify_peer_func)(const char* target_host,
+                                                      const char* target_pem,
+                                                      void* userdata,
+                                                      int32_t isDestroy);
+
+static void grpcsharp_verify_peer_destruct_handler(void* userdata) {
```
Comment (jtattermusch) [inline src/csharp/ext/grpc_csharp_ext.c]: nit: replace "destruct" with "destroy"

**Comment by jtattermusch** [comment]
Test failures:
https://source.cloud.google.com/results/invocations/0d7774de-3444-4b25-b56e-adef8ba14479/targets/github%2Fgrpc%2Fcsharp_linux_dbg_native/tests (a problem with the logic?)

A few problems with sanity tests:
https://source.cloud.google.com/results/invocations/1e949f76-9b3d-412a-9a6e-3c777ded5863/targets/github%2Fgrpc%2Fsanity_linux_dbg_native/tests
(formatting is something we can solve as the last step).

File: src/csharp/Grpc.IntegrationTesting/SslCredentialsTest.cs (hunk 15 above)
Comment (tlg-bf) [reply src/csharp/Grpc.IntegrationTesting/SslCredentialsTest.cs]: The `certChain` obtained from the file contains `\r`, but in the callback, the string passed by the unmanaged code doesn't contain `\r`, which lead the equality line 246 to fail. I guess I could strip the `\r` in the callback, let me know if you think it is better.

File: src/csharp/Grpc.Core/ChannelCredentials.cs (hunk 16 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/ChannelCredentials.cs]: Thanks, replaced by "false".

File: src/csharp/Grpc.Core/Internal/NativeMethods.Generated.cs (hunk 19 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/Internal/NativeMethods.Generated.cs]: I ran the script, it generated quite many changes but I assumed that only the one for this files were relevant.

File: src/csharp/Grpc.Core/ChannelCredentials.cs (hunk 18 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/ChannelCredentials.cs]: Thanks I fixed it.

File: src/csharp/Grpc.Core/ChannelCredentials.cs (hunk 17 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/ChannelCredentials.cs]: Renamed

File: src/csharp/ext/grpc_csharp_ext.c (hunk 20 above)
Comment (tlg-bf) [reply src/csharp/ext/grpc_csharp_ext.c]: Renamed

File: src/csharp/Grpc.Core/Channel.cs (hunk 11 above)
Comment (tlg-bf) [reply src/csharp/Grpc.Core/Channel.cs]: Thanks for the thorough answer!

**Comment by tlg-bf** [comment]
> Test failures:
> https://source.cloud.google.com/results/invocations/0d7774de-3444-4b25-b56e-adef8ba14479/targets/github%2Fgrpc%2Fcsharp_linux_dbg_native/tests (a problem with the logic?)

I haven't managed to reproduce the error for now on windows or in wsl. Do you know how I could try to reproduce this locally?

Looking at the log it seems to be the destroy callback throwing a null pointer from the managed code at some point, but without being able to reproduce it is difficult to determine precisely where or why the exception is happening.

Looking at the code, the only thing that could throw a null pointer exception in the `VerifyPeerCallbackHandler` would be the `this.verifyPeerCallback(...)` line, but I don't really see how that could be happening.

**Comment by tlg-bf** [comment]
Ok I finally managed to reproduce the bug using a Linux container and Mono. So it seems to only show up when running the test in Mono, and what is happening is that the `VerifyPeerHandler` callback is invoked by the managed code, but the object itself seem to have been garbage collected despite the `gcHandle` (probably the CLR forcibly collects it when shutting down), so a `NullReferenceException` is thrown when calling `this.gcHandle.Free()`.

I think the issue can be solved in two ways:
* Put a try catch around the `this.gcHandle.Free()` and just ignore if it throws an exception. I put this solution on a [separate branch](https://github.com/tlg-bf/grpc/tree/2-provide-access-to-client-certificate-quick-fix), and it looks like this:

https://github.com/tlg-bf/grpc/blob/57acf5eabaf0cbcf6b0b21b10e22bb9b55c00669/src/csharp/Grpc.Core/ChannelCredentials.cs#L215-L231

* Use a finalizer to make sure that the object is not disposed until the unmanaged code has invoked the callback with the `isDestroy` true. Relevant pieces of codes for [this branch](https://github.com/tlg-bf/grpc/tree/2-provide-access-to-client-certificate-finalize) are:

https://github.com/tlg-bf/grpc/blob/693da6ae8be81b415161fe41a017fcc76a32b393/src/csharp/Grpc.Core/ChannelCredentials.cs#L173-L183

https://github.com/tlg-bf/grpc/blob/693da6ae8be81b415161fe41a017fcc76a32b393/src/csharp/Grpc.Core/ChannelCredentials.cs#L186-L196

https://github.com/tlg-bf/grpc/blob/693da6ae8be81b415161fe41a017fcc76a32b393/src/csharp/Grpc.Core/ChannelCredentials.cs#L232-L239

Let me know what you think is the best way or if you have a better solution.

**Comment by jtattermusch** [comment]
It looks like that getting this right is pretty complicated - so I'm thinking of adding some utility code, that would allow others to use the native callbacks without needing to replicate the complex logic around native object lifetime.  At the same time, I'm hoping the newly added code would solve https://github.com/grpc/grpc/issues/18178 and https://github.com/grpc/grpc/issues/17587. Please stay tuned.

**Comment by georghinkel** [comment]
Given that #18327 is merged now, does this have a chance to be merged soon?

**Comment by tlg-bf** [comment]
I updated to use the new callback registration mechanism trying to follow how it was done in `NativeMetadataCredentialsPlugin` . @jtattermusch let me know if there is anything wrong.

**Comment by jtattermusch** [comment]
> I updated to use the new callback registration mechanism trying to follow how it was done in `NativeMetadataCredentialsPlugin` . @jtattermusch let me know if there is anything wrong.

Thanks! In the meantime, I was working on integration with NativeCallbackDispatcher myself. The resulting code is similar (I started off of your changes), but I did some extra cleanup and rewrote the tests (they weren't setup in a very readable way IMHO). Also your PR still has some bug I believe. Let's review https://github.com/grpc/grpc/pull/18591 instead and close this PR as superseded.
